homeassistant/components/firmata/__init__.py

```diff
@@ -0,0 +1,170 @@
+from .board import FirmataBoard
+from .const import (
+    CONF_ARDUINO_INSTANCE_ID,
+    CONF_ARDUINO_WAIT,
+    CONF_BINARY_SENSORS,
+    CONF_INITIAL_STATE,
+    CONF_NEGATE_STATE,
+    CONF_PIN,
+    CONF_PIN_MODE,
+    CONF_SAMPLING_INTERVAL,
+    CONF_SERIAL_BAUD_RATE,
+    CONF_SERIAL_PORT,
+    CONF_SLEEP_TUNE,
+    CONF_SWITCHES,
+    DOMAIN,
+    PIN_MODE_INPUT,
+    PIN_MODE_OUTPUT,
+    PIN_MODE_PULLUP,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+DATA_CONFIGS = "board_configs"
+
+ANALOG_PIN_SCHEMA = vol.All(cv.string, vol.Match(r"^A[0-9]+$"))
+
+SWITCH_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_NAME): cv.string,
+        vol.Required(CONF_PIN): vol.Any(cv.positive_int, ANALOG_PIN_SCHEMA),
+        # will be analog mode in future too
+        vol.Required(CONF_PIN_MODE): PIN_MODE_OUTPUT,
+        vol.Optional(CONF_INITIAL_STATE, default=False): cv.boolean,
+        vol.Optional(CONF_NEGATE_STATE, default=False): cv.boolean,
+    },
+    required=True,
+    extra=vol.ALLOW_EXTRA,
+)
+
+BINARY_SENSOR_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_NAME): cv.string,
+        vol.Required(CONF_PIN): vol.Any(cv.positive_int, ANALOG_PIN_SCHEMA),
+        # will be analog mode in future too
+        vol.Required(CONF_PIN_MODE): vol.Any(PIN_MODE_INPUT, PIN_MODE_PULLUP),
+        vol.Optional(CONF_NEGATE_STATE, default=False): cv.boolean,
+    },
+    required=True,
+    extra=vol.ALLOW_EXTRA,
+)
+
+BOARD_CONFIG_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_SERIAL_PORT): cv.string,
+        vol.Optional(CONF_SERIAL_BAUD_RATE): cv.positive_int,
+        vol.Optional(CONF_ARDUINO_INSTANCE_ID): cv.positive_int,
+        vol.Optional(CONF_ARDUINO_WAIT): cv.positive_int,
+        vol.Optional(CONF_SLEEP_TUNE): vol.All(
+            vol.Coerce(float), vol.Range(min=0.0001)
+        ),
+        vol.Optional(CONF_SAMPLING_INTERVAL): cv.positive_int,
+        vol.Optional(CONF_SWITCHES): [SWITCH_SCHEMA],
+        vol.Optional(CONF_BINARY_SENSORS): [BINARY_SENSOR_SCHEMA],
+    },
+    required=True,
+    extra=vol.ALLOW_EXTRA,
+)
+
+CONFIG_SCHEMA = vol.Schema(
+    {DOMAIN: vol.All(cv.ensure_list, [BOARD_CONFIG_SCHEMA])}, extra=vol.ALLOW_EXTRA
+)
+
+
+async def async_setup(hass, config):
+    """Set up the Firmata domain."""
+    # Delete specific entries that no longer exist in the config
+    if hass.config_entries.async_entries(DOMAIN):
+        for entry in hass.config_entries.async_entries(DOMAIN):
+            remove = True
+            for board in config[DOMAIN]:
+                if entry.data[CONF_SERIAL_PORT] == board[CONF_SERIAL_PORT]:
+                    remove = False
+                    break
+            if remove:
+                await hass.config_entries.async_remove(entry.entry_id)
+
+    # Setup new entries and update old entries
+    for board in config[DOMAIN]:
+        firmata_config = copy(board)
+        existing_entry = False
+        for entry in hass.config_entries.async_entries(DOMAIN):
```

Comment:
Looks like we could merge the update and delete loops since we're iterating the same things. Just do the operations in the correct order and it should work.

Reply:
On second thought, maybe that would change an iterable during iteration which is an error?

Reply:
When I was programming I attempted to merge them, but it made it messy. If you loop with `hass.config_entries.async_entries(DOMAIN)` as the parent, then you still need a way to loop through the remaining entries in `config[DOMAIN]`, and if you loop `config[DOMAIN]` as the parent, then you still need a way to loop through the remaining entries in `hass.config_entries.async_entries(DOMAIN)`. The only way I could do this was to mark some as used with another list, but that still required another loop after that process to find the unused entries.

Reply:
Ok, let's keep it like this for now.